src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1228 @@
+        );
+        assert_eq!(persistent_store.min_pin_length(), min_pin_length);
+    }
+
+    #[test]
+    fn test_process() {
+        let mut rng = ThreadRng256 {};
+        let mut persistent_store = PersistentStore::new(&mut rng);
+        let mut pin_protocol_v1 = PinProtocolV1::new(&mut rng);
+        let client_pin_params = AuthenticatorClientPinParameters {
+            pin_protocol: 1,
+            sub_command: ClientPinSubCommand::GetPinRetries,
+            key_agreement: None,
+            pin_auth: None,
+            new_pin_enc: None,
+            pin_hash_enc: None,
+            #[cfg(feature = "with_ctap2_1")]
+            min_pin_length: None,
+            #[cfg(feature = "with_ctap2_1")]
+            min_pin_length_rp_ids: None,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions: None,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions_rp_id: None,
+        };
+        assert!(pin_protocol_v1
+            .process_subcommand(&mut rng, &mut persistent_store, client_pin_params)
+            .is_ok());
+
+        let client_pin_params = AuthenticatorClientPinParameters {
+            pin_protocol: 2,
+            sub_command: ClientPinSubCommand::GetPinRetries,
+            key_agreement: None,
+            pin_auth: None,
+            new_pin_enc: None,
+            pin_hash_enc: None,
+            #[cfg(feature = "with_ctap2_1")]
+            min_pin_length: None,
+            #[cfg(feature = "with_ctap2_1")]
+            min_pin_length_rp_ids: None,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions: None,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions_rp_id: None,
+        };
+        #[cfg(not(feature = "with_ctap2_1"))]
+        let error_code = Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID;
+        #[cfg(feature = "with_ctap2_1")]
+        let error_code = Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER;
+        assert_eq!(
+            pin_protocol_v1.process_subcommand(&mut rng, &mut persistent_store, client_pin_params),
+            Err(error_code)
+        );
+    }
+
+    #[test]
+    fn test_decrypt_pin() {
+        let shared_secret = [0x88; 32];
+        let aes_enc_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let aes_dec_key = crypto::aes256::DecryptionKey::new(&aes_enc_key);
+
+        // "1234"
+        let new_pin_enc = vec![
+            0xC0, 0xCF, 0xAE, 0x4C, 0x79, 0x56, 0x87, 0x99, 0xE5, 0x83, 0x4F, 0xE6, 0x4D, 0xFE,
+            0x53, 0x32, 0x36, 0x0D, 0xF9, 0x1E, 0x47, 0x66, 0x10, 0x5C, 0x63, 0x30, 0x1D, 0xCC,
+            0x00, 0x09, 0x91, 0xA4, 0x20, 0x6B, 0x78, 0x10, 0xFE, 0xC6, 0x2E, 0x7E, 0x75, 0x14,
+            0xEE, 0x01, 0x99, 0x6C, 0xD7, 0xE5, 0x2B, 0xA5, 0x7A, 0x5A, 0xE1, 0xEC, 0x69, 0x31,
+            0x18, 0x35, 0x06, 0x66, 0x97, 0x84, 0x68, 0xC2,
+        ];
+        assert_eq!(
+            decrypt_pin(&aes_dec_key, new_pin_enc),
+            Some(vec![0x31, 0x32, 0x33, 0x34]),
```

Comment:
Nit: `vec![b'1', b'2', b'3', b'4']` would be more readable. Or even `b"1234".into_vec()` (if that doesn't include any trailing zero). Same for the other tests.

Reply:
Changed both tests in this function.

Reply:
I still see the "manual vec! form" in the following test function.

Reply:
That is because I need a `0u8` there, and that was the most consistent and visible solution to me.

Reply:
On the other hand, the comment above the test mentions "12'\0'4", so why not using `b"12\04"` there?

Reply:
I wasn't sure if it's too easy to miss, but done.